### PR TITLE
Fix #8860.

### LINKED ISSUE #8860: [GCStress=C] Assert failure(PID 20160 [0x00004ec0], Thread: 10876 [0x2a7c]): !CREATE_CHECK_STRING(!"Detected use of a corrupted OBJECTREF. Possible GC hole.")
https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/x64_checked_windows_nt_gcstress0xc/13/testReport/

**Configuration**
Windows x64 Checked
COMPlus_GCStress=C

**Affected tests**
JIT\SIMD\StoreElement_r
JIT\SIMD\BoxUnbox_r
JIT\SIMD\AbsGeneric_ro
JIT\SIMD\AbsGeneric_r

**Output**
```
Assert failure(PID 16168 [0x00003f28], Thread: 12960 [0x32a0]): !CREATE_CHECK_STRING(!"Detected use of a corrupted OBJECTREF. Possible GC hole.")

CORECLR! `Object::ValidateInner'::`1'::catch$12 + 0x151 (0x00007ffa`a0a5febf)
CORECLR! CallSettingFrame + 0x20 (0x00007ffa`a094ec30)
CORECLR! _CxxCallCatchBlock + 0x15A (0x00007ffa`a094abba)
NTDLL! RtlCaptureContext + 0x3C3 (0x00007ffa`e550a193)
CORECLR! Object::ValidateInner + 0x15C (0x00007ffa`9ff8d2bc)
CORECLR! Object::Validate + 0x132 (0x00007ffa`9ff8cdc2)
CORECLR! GcInfoDecoder::ReportRegisterToGC + 0x123 (0x00007ffa`a06e14a3)
CORECLR! GcInfoDecoder::ReportSlotToGC + 0xCD (0x00007ffa`a06e16d5)
CORECLR! GcInfoDecoder::EnumerateLiveSlots + 0xDE0 (0x00007ffa`a06e08ac)
CORECLR! EECodeManager::EnumGcRefs + 0x383 (0x00007ffa`a038a1d3)
    File: c:\git\coreclr\src\vm\object.cpp Line: 1827
    Image: C:\git\coreclr\bin\tests\Windows_NT.x64.Checked\Tests\Core_Root\CoreRun.exe

Expected: 100
Actual: 123456789
END EXECUTION - FAILED
FAILED
```

### PR DESCRIPTION
**Pull request by pgavlin**
These failure was an assertion under GC stress due to registers that
were reported as live while containing an invalid GC reference. The root
cause of this assertion was the use of SIMD instructions that did not
kill any GC ref contained in their destination register. This change fixes
the bug by properly killing any GC ref in the destination register for
such an instruction, which required two changes:
- Instructions with the format `IF_RRW_RRW_CNS` were never killing the
  destination register. The code that handles this format has been
  updated to do so when necessary.
- `pextrw` and `pmovmskb` were not handled correctly in
  `emitInsCanOnlyWriteSSE2OrAVXReg`. These instructions may write to
  GPRs, so that function must return false when they are passed as
  input.

### CODE REVIEW

**Comment by pgavlin** [comment]
@sivarv @CarolEidt @dotnet/jit-contrib PTAL

Fixes part of #8860.

**Comment by sivarv** [comment]
:shipit:

**Comment by sivarv** [comment]
@briansul - this is a gc-hole fix we would need even on desktop 4.7.

File: src/jit/emitxarch.cpp [hunk 1]
```diff
@@ -10832,6 +10832,9 @@ size_t emitter::emitOutputInstr(insGroup* ig, instrDesc* id, BYTE** dp)
 
             dst += emitOutputByte(dst, emitGetInsSC(id));
             sz = emitSizeOfInsDsc(id);
+
+            // Kill any GC ref in the destination register.
+            emitGCregDeadUpd(id->idReg1(), dst);
```
Comment (sivarv) [inline src/jit/emitxarch.cpp]: On second further thinking, we can avoid call overhead for those instructions that write to an sse/avx reg by conditioning this on emitInsCanOnlyWriteSSE2OrAVXReg()?

e.g. if (emitInsCanOnlyWriteSSE2OrAVXReg())
       {
           // do nothing
       } 
      else
       {
             emitGCregDeadUpd(id->idReg1(), dst);
      }

I see emitInsCanOnlyWriteSSE2OrAVXReg() needs to be updated to handle pextrw and pmovkskb instructions.

Comment (pgavlin) [inline src/jit/emitxarch.cpp]: > we can avoid call overhead for those instructions that write to an sse/avx reg by conditioning this on 

Seems reasonable.

> I see emitInsCanOnlyWriteSSE2OrAVXReg() needs to be updated to handle pextrw and pmovkskb instructions.

This is the fix for at least one other piece of #8860. I'm happy to fold it into this change as well.

**Comment by CarolEidt** [comment]
This looks OK; however I thought it was the case that we should be marking GC refs as dead as soon as they are no longer in use. Have you looked into why that was not the case for this?

**Comment by pgavlin** [comment]
@CarolEidt: that is the case during codegen, but not during emit. The emit phase relies on recognizing changes to the contents of registers in order to update register liveness.

**Comment by CarolEidt** [comment]
@pgavlin - thanks for clarifying.
LGTM

**Comment by sivarv** [comment]
Looks good.

**Pull request merged**